### PR TITLE
Prohibit starting of postmaster with EXECUTE role.

### PR DESCRIPTION
**Pull request by dkovalev1**
Prohibit starting of postmaster with EXECUTE role.

Normally segment node is started in utility role and execute role is used only
when backend process is started. However if one tries to start segment directly
in the execute mode using pg_ctl, this leads to SIGSEGV in the postgres process.

It happens because pointer to shared memory flag shmCleanupBackends is left
uninitialized. This pointer initialized in the function tmShmemInit() only for
GP_ROLE_DISPATCH and GP_ROLE_UTILITY and the flag itself is meaningful only for
GP_ROLE_DISPATCH.

This patch prohibits start of the postmaster in execute mode and raises error in
this case. To ensure that flag shmCleanupBackends is initialized correctly an
assert was added.

This scenario assumes unexpected startup procedure and quite specific use case
not falling into regular regression tests. Automated testing of this case would
require quite complicated TAP test and does not bring any value to normal use
cases. Considering this, it seems not feasible to make a specific automated test
for this scenario.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/81914

File: src/backend/access/transam/xlog.c [hunk 1]
```diff
@@ -8207,7 +8207,8 @@ StartupXLOG(void)
 	if (fast_promoted)
 		RequestCheckpoint(CHECKPOINT_FORCE);
 
-	*shmCleanupBackends = true;
+	 if (Gp_role == GP_ROLE_DISPATCH)
+		*shmCleanupBackends = true;
```
Comment (RekGRpth) [inline src/backend/access/transam/xlog.c]: do we need [this](https://github.com/arenadata/gpdb/commit/b04175666aefe49f7410ff010d84998d86d20d4c#diff-c1cb3ab2a19606390c1a7ed00ffe5a45531702ca5faf999d401c548f8951c65bR6294-R6296) change too?

Comment (dkovalev1) [reply src/backend/access/transam/xlog.c]: Do you mean to merge the whole patch "Try to terminate mpp backends during dtx recovery." or just this particular point of triggering restart? 
7.x has reworked restart procedure, I have doubts it's good idea to backport the whole patch.

Comment (RekGRpth) [reply src/backend/access/transam/xlog.c]: maybe, I don't remember why I didn't port this place.

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/81995

**Comment by KnightMurloc** [comment]
What does it mean to run postmaster (if I understood correctly) with the execute role? Is there any point in supporting this?

**Comment by dkovalev1** [comment]
> What does it mean to run postmaster (if I understood correctly) with the execute role? Is there any point in supporting this?

If we try to start it with pg_ctl  with f.e. -o "gp_role=execute", this will try to set gp_role to execute at postmaster startup. This is normal for adb 7.x, but makes no sense for 6.x.
Execute role (QE) is normal operation mode for the backend on the segments and set only when user session starts. What do you mean by supporting this?

File: src/backend/access/transam/xlog.c [hunk 2]
```diff
@@ -8207,7 +8211,8 @@ StartupXLOG(void)
 	if (fast_promoted)
 		RequestCheckpoint(CHECKPOINT_FORCE);
 
-	*shmCleanupBackends = true;
```
Comment (RekGRpth) [inline src/backend/access/transam/xlog.c:8210]: Maybe add some words to the description about how we can reach this place with `shmCleanupBackends` equal to `NULL`?
And similarly about the place above.

**Comment by KnightMurloc** [comment]
> What do you mean by supporting this?

I mean, what kind of user case is there for this? Might it be worth adding a check and fail with an meaningful error if postmaster is started with this role?

**Comment by KnightMurloc** [comment]
It seems to me that it is worth prohibiting the start of postmaster with the execute role, since this does not make practical sense and if for some reason it is started with this role, then this is clearly an error.

File: src/backend/access/transam/xlog.c [hunk 3]
```diff
@@ -8207,7 +8211,8 @@ StartupXLOG(void)
...
+	 if (Gp_role == GP_ROLE_DISPATCH)
```
Comment (KnightMurloc) [inline src/backend/access/transam/xlog.c]: Remove the extra space.
```suggestion
	if (Gp_role == GP_ROLE_DISPATCH)
```

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/82221

**Comment by dkovalev1** [comment]
> It seems to me that it is worth prohibiting the start of postmaster with the execute role, since this does not make practical sense and if for some reason it is started with this role, then this is clearly an error.

Although it seems good idea for me as well, it's not very practical to do it. Here we don't start the it in the specific mode. Instead it just started normally and pass a set of options to it, one of it (okay, may be the one we pass), by coincidence is gp_role mode.
When postgres process starts it applies a set of options from defaults, from configuration file, command line, and some of that option combinations might be invalid, incorrect or differ from the user's expectations. AFAIK the is no and it's extremely difficult to check all options combinations for correctness, this is left for the user - most likely developer or DBA in this case.
Also, the user don't start pg_ctl directly in case of GP/ADB, instead he using scripts which already contain correct options set. And, I am not completely sure that starting with gp_role=execute has no practical usage.
So, I believe checking for gp_role=execute at start is not feasible.

**Comment by KnightMurloc** [comment]
However if we know that one of the options is incorrect, then it is worth informing the user about it as early as possible. Moreover, in this state, we cannot even take a lock on the table.

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/82607

**Comment by dkovalev1** [comment]
> However if we know that one of the options is incorrect, then it is worth informing the user about it as early as possible. Moreover, in this state, we cannot even take a lock on the table.

Okay

File: src/backend/access/transam/xlog.c [hunk 4]
```diff
@@ -8207,7 +8211,8 @@ StartupXLOG(void)
 	if (fast_promoted)
 		RequestCheckpoint(CHECKPOINT_FORCE);
 
-	*shmCleanupBackends = true;
+	if (Gp_role == GP_ROLE_DISPATCH)
+		*shmCleanupBackends = true;
```
Comment (KnightMurloc) [inline src/backend/access/transam/xlog.c]: Maybe it's worth replacing the check with an Assert that the pointer is not NULL?

Comment (dkovalev1) [reply src/backend/access/transam/xlog.c]: Assert works only in debug build.

Comment (KnightMurloc) [reply src/backend/access/transam/xlog.c]: That's the point. If we now assume that this cannot happen, what is the point of checking this in the release build?

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/82618

File: src/backend/access/transam/xlog.c [hunk 5]
```diff
@@ -6688,6 +6688,10 @@ StartupXLOG(void)
 		 */
 		if (access(BACKUP_LABEL_FILE, F_OK) != 0)
 				SyncAllXLogFiles();
+				
+		Assert(shmCleanupBackends != NULL);
+		*shmCleanupBackends = true;
```
Comment (KnightMurloc) [inline src/backend/access/transam/xlog.c]: This has not been here before. Do we need it?

File: src/backend/access/transam/xlog.c (hunk 2 above)
Comment (dkovalev1) [reply src/backend/access/transam/xlog.c:8210]: added

File: src/backend/access/transam/xlog.c (hunk 5 above)
Comment (dkovalev1) [reply src/backend/access/transam/xlog.c]: FYI: https://github.com/arenadata/gpdb/pull/1065#discussion_r1783907736

Comment (KnightMurloc) [reply src/backend/access/transam/xlog.c]: I looked at it and I don't really see the meaning of this code. In any case, we will set this variable to true at the end of the function. Is there a situation when we exit this function earlier (except for errors)? Could we need this value earlier?

Comment (dkovalev1) [reply src/backend/access/transam/xlog.c]: There are errors possible, of different severity, this is the way to exit. Called functions are also capable to throw an error. In this case we might want to exit with this flag set, which means that MPP backends should be terminated. IMHO it's better to have this flag as in 7.x than not. @RekGRpth , what do you think?

Comment (RekGRpth) [reply src/backend/access/transam/xlog.c]: I think to prove the need for this change - we need to come up with a scenario when it is needed. Like in the [patch](https://github.com/arenadata/gpdb/pull/567), a scenario was come up for a similar change at the end of the function.

Comment (dkovalev1) [reply src/backend/access/transam/xlog.c]: Okay, let's consider that if this change is needed it's a matter of a different issue/PR.

File: src/backend/access/transam/xlog.c [hunk 6]
```diff
@@ -6688,6 +6688,8 @@ StartupXLOG(void)
 		 */
 		if (access(BACKUP_LABEL_FILE, F_OK) != 0)
 				SyncAllXLogFiles();
+				
+
```
Comment (KnightMurloc) [inline src/backend/access/transam/xlog.c]: Remove extra lines.

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/82874

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/82880

**Review by KnightMurloc — DISMISSED** [review]

File: src/backend/postmaster/postmaster.c [hunk 7]
```diff
@@ -1055,6 +1055,20 @@ PostmasterMain(int argc, char *argv[])
...
+		 * Starting it in EXECUTE mode is impossible because it skips several 
+		 * initialization stages and becomes unusable.
+		 * It never started in EXECUTE mode when from supporting scripts,
+		 * but it's still possible to try it with pg_ctl manually. To inform 
+		 * curious folks who do it, check and bail out.
+		 */
+	    ereport(FATAL,
+            (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+             errmsg("Postmaster can not be started in EXECUTE mode.")));
+	}
```
Comment (RekGRpth) [inline src/backend/postmaster/postmaster.c]: Could you please pgindent your changes?

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/83097

File: src/backend/postmaster/postmaster.c [hunk 8]
```diff
@@ -1055,6 +1055,21 @@ PostmasterMain(int argc, char *argv[])
 		ereport(ERROR,
 				(errmsg("WAL streaming (max_wal_senders > 0) requires wal_level \"archive\", \"hot_standby\", or \"logical\"")));
 
+	if (Gp_role == GP_ROLE_EXECUTE)
+	{
+		/**
+		 * Postmaster normally started in DISPATCH or UTILITY mode and switched to EXECUTE afterwards.
```
Comment (RekGRpth) [inline src/backend/postmaster/postmaster.c]: I think this line is still too long.

**Review by RekGRpth — APPROVED** [review]

**Review by KnightMurloc — APPROVED** [review]

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/83120

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/83509

**Pull request merged**